### PR TITLE
Continuous bernoulli distribution

### PR DESCRIPTION
**Pull request by gabloa**
We recently had a NeurIPS paper (https://arxiv.org/abs/1907.06845 and https://papers.nips.cc/paper/9484-the-continuous-bernoulli-fixing-a-pervasive-error-in-variational-autoencoders) where we introduce a new (0,1)-supported distribution: the continuous Bernoulli. This pull request implements this distribution in pytorch.

### CODE REVIEW

**Comment by mruberry** [comment]
@fritzo, are you the correct reviewer for this PR? If not, just let me know.

**Comment by fritzo** [comment]
cc @neerajprad @alicanb 

@mruberry yes one of us should have time to review in the next few days.

**Comment by mruberry** [comment]
Thanks @fritzo!

File: torch/distributions/constraints.py
Comment (fritzo) [inline torch/distributions/constraints.py]: I'd rather avoid introducing the notion of open/closed constraints in this library, since most of the algorithms cannot distinguish open vs closed, and introducing the distinction could lead to a proliferation of incompatible types. Do you have a specific reason for using `open_interval` rather than the existing `interval`?

**Review by fritzo — CHANGES_REQUESTED** [review]
Looks thorough on first pass! I have not verified math. Two main comments:

1. You'll need to add tests to test/test_distributions.py. You can follow an existing distribution for example. I would recommend that, when adding new test examples to existing lists, you append new examples to lists rather than  inserting them in alphabetical order; this reduces random seed churn which triggers flaky tests. You can also check your custom `.entropy()` method against that provided by `ExponentialFamily`.

2. Regarding open/closed intervals, I'd recommend using the existing `interval` and clamping inputs as is done in other distributions (see distributions/utils.py for example). Adding a new constraint requires significant design choices and I suspect is out of scope for this PR (but I'm open to starting a design discussion with other module maintainers).

File: torch/distributions/constraints.py
Comment (gabloa) [reply torch/distributions/constraints.py]: Thanks for the input! I'll write the tests and remove the custom entropy method to rely on the one from ExponentialFamily. As for the open interval, the issue I was trying to avoid is that the distribution is not properly defined when probs=0 or probs=1 (as evidenced by all the log(probs) and log(1-probs) terms appearing throughout the code). However clamping should work just fine, I will also change this.

**Comment by gabloa** [comment]
Pushed again including your comments. I ended up leaving the custom entropy method as it was faster and doesn't require computing more gradients internally, and actually some exponential family distributions also have their own entropy method.

File: torch/distributions/continuous_bernoulli.py
Comment (fritzo) [inline torch/distributions/continuous_bernoulli.py]: nit: can you make `lims` a tuple so it is immutable? In general it is safer to make all Python arg defaults immutable.

File: torch/distributions/kl.py
Comment (fritzo) [inline torch/distributions/kl.py]: nit: can you replace this with a `torch.where()`? I think some of the older kl implementations use in-place ops and should eventually also use `torch.where()`.

**Review by fritzo — CHANGES_REQUESTED** [review]
Thanks for your patience @gabloa. Implementation looks thorough and tests look good. I have only a couple comments. Could you also merge master before pushing the changes? Current test failures appear unrelated.

Is there any part of the may you'd like me to pay special attention to while reviewing?

File: test/test_distributions.py
Comment (gabloa) [reply test/test_distributions.py]: Changed this to include a function which explicitly computes the values instead of hard-coding them.

**Comment by gabloa** [comment]
Pushed the new updates after merging with master, I don't know what I did wrong but github automatically requested reviews from everyone and is now displaying basically all files as changed, apologies about this. Please let me know if you'd like me to do something on my side to fix this.

**Comment by fritzo** [comment]
Thanks for updating the tests, that looks much easier to maintain!

> I pushed the new updates after merging with master, I don't know what I did wrong

Yikes, I don't know what happened either. Could you try merging master again? If all else fails you can make a backup branch, `git reset --hard master`, and manually copy the few files you touched from the backup branch. Every once in a while I have to hard reset myself out of such a mess 🤷‍♂

**Comment by albanD** [comment]
For the master issue, I think it is mainly that it does not detect your master merge as being part of master.
You can:
- Remove the merge commit and then rebase your branch on top of master.
- Create a brand new branch based on current master and cherry-pick your commits into it.
- As @fritzo proposed: create a branch new branch and redo your changes.

**Comment by gabloa** [comment]
I tried rebasing master as @albanD suggested but github is still detecting changes in pytorch master after the date when I forked pytorch as changes in this PR... Also, additional reviewers that were added automatically have not been removed (apologies to everyone whose email I'm spamming). Maybe the easiest fix is for me to just fork pytorch again, create a new branch there with my updates, make a new PR and close this one?

**Comment by ezyang** [comment]
Yeah, go ahead and just make a new PR

**Comment by gabloa** [comment]
Ok, just opened a new pull request (#34619). Thanks and again, sorry about the mess!

**Pull request closed without merging**
